README
			=========================
			PREDICT QUICK START GUIDE
			=========================

SYSTEM REQUIREMENTS
===================
This program requires ncurses version 4.2 or higher.  Earlier versions have
been known to cause segmentation faults and/or odd display colors with this
program.  ncurses may be obtained via:

	https://www.gnu.org/software/ncurses/


The Linux pthreads library is also required.  Both of these libraries are
usually available in modern Linux distributions.


UNPACKING
=========
The PREDICT archive can be unpacked in any directory for which read access
is provided for all system users (such as /usr/src or /usr/local/src for
Slackware Linux).  As root, move the tar file to such a directory and
issue the command:

	tar xvfz predict-2.2.4.tar.gz

to unpack the file.


COMPILATION
===========
PREDICT's compilation and installation procedure differs slightly from
that of most software, but should easy enough to understand and follow.
First, move (cd) into the predict directory and execute the included
"configure" script by typing:

	./configure

as root at your command prompt.  This script compiles and runs the install
program that configures the source code for PREDICT.  It then compiles and
installs PREDICT by creating symbolic links between the executables created
in the installation directory and /usr/local/bin.  If a destination directory
other than /usr/local/bin is desired, invoke configure with the desired
directory as an argument on the command line:

	./configure /usr/bin

NOTE: "configure" may stall if the system soundcard is in use.  Some
window managers (sometimes KDE) take control of the soundcard, and prevent
the configure script from opening /dev/dsp and checking its existence. 


FIRST TIME USE
==============
First time users will be asked to enter their groundstation latitude and
longitude in decimal degrees or in degree, minute, second (DMS) format.
Normally, PREDICT handles longitudes in decimal degrees WEST (0-360
degrees), and latitudes in decimal degrees NORTH.  This behavior can be
modified by passing the -east or -south command line switches to PREDICT
when it is invoked.  Your station's altitude in meters above mean sea
level, a recent set of Keplerian orbital data for the satellites of
interest, and an accurately set system clock are also required if
successful real-time satellite tracking is also desired.  Sources
for such data include http://www.celestrak.com/,
http://www.space-track.org, and http://www.amsat.org/.


FOR FURTHER INFORMATION
=======================
Please consult the files under the "docs" subdirectory for more
directions on the use and capabilities of PREDICT.

The latest news and information regarding PREDICT software is
available at: http://www.qsl.net/kd2bd/predict.html.


Happy Tracking!
 
--
John, KD2BD
kd2bd@amsat.org
March 2018



NEWS
PREDICT 2.2.0 Release
=====================
The major revisions in PREDICT 2.2.0 include the use of
SGP4/SDP4 orbital models for improved tracking accuracy,
as well as a re-written Single Satellite Tracking Mode
to provide more useful real-time satellite communications
data.


New Orbital Models
==================
The simplified orbital model used in previous versions
of PREDICT provided very accurate and dependable results
for the majority of users, but was not quite good enough
for more demanding applications at NASA.

In July 2002, a test was performed on the accuracy of the
orbital model used in PREDICT.  The test was performed by
a software engineer working at Edwards Airforce Base in
California, and involved comparing data provided by PREDICT's
Calc() and PreCalc() functions to that of a radar system
used to track the International Space Station.  Ranging
data from the radar was taken every 50 milliseconds for
the duration of the pass.

The error in azimuth was found to be almost zero at AOS, and
increased to 0.3 degrees over the duration of the pass.  The
elevation was about 0.2 degrees lower than that measured by
radar for the entire pass.  The calculated range was approximately
7000 meters low at AOS, and approximately 7000 meters high at
LOS with a crossover at TCA.


New Single Satellite Tracking Mode
==================================
The Single Satellite Tracking Mode has been completely revised
in PREDICT version 2.2.0.  PREDICT now supports a database of
satellite transponder data so that the program can provide
path loss and Doppler-corrected uplink and downlink frequency
information in real-time.  For satellites containing linear
transponders, uplink and downlink frequencies may be adjusted
by pressing the '<' and '>' keys to tune down across the
transponder in 1 kHz steps, or the ',' and '.' keys (unshifted
'<' and '>') to tune in 100 Hz steps.  Switching between
transponders may be accomplished by pressing the SPACE bar.

Eclipse depth, an angle representing how deep a satellite is
in solar eclipse, is also provided.  Eclipse depth is positive
when the spacecraft is in eclipse, and negative when it is in
sunlight.  Spacecraft antenna squint angle is provided if
ALAT/ALON (spacecraft attitude coordinates) parameters are
included in the transponder database file.


Transponder Database
====================
When PREDICT version 2.2.0 is initially run, a default transponder
database file (predict.db) is copied into the user's $HOME/.predict
directory.  At present, the database provides for many new features,
som of which have not been completely implemented by PREDICT in
this early release of the 2.2.x run.  The most glaring omission
is lack of a facility to edit or modify the database file.  For
the time being, it may be edited by hand if care is excercised.

The format for the database file is as follows.  Using NOAA-17
as an example:

NOAA-17			Name of satellite
27453			Object number (must match predict.tle)
No alat/alon		ALAT/ALON spacecraft attitude information
APT Downlink		Description of transmitter/transponder
0.0, 0.0		Uplink passband limits in MHz (0.0 = no uplink)
137.620, 137.620	Corresponding downlink passband limits (single downlink)
No weekly schedule	Days of the week this transponder is active
No orbital schedule	Orbital phase when this transponder is active
HRPT Downlink		Description of second transmitter/transponder
0.0, 0.0		Uplink passband limits (low end limit, high end)
1707.0, 1707.0		Corresponding downlink passband limits (single downlink)
No weekly schedule	Days of the week this transponder is active
No orbital schedule	Orbital phase when this transponder is active
end			End of data for this satellite

OSCAR-29 contains linear (inverting) as well as digital (Pacsat)
transponders:

OSCAR-29
24278
No alat/alon
Mode JA Transponder
145.900, 146.000
435.900, 435.800
No weekly schedule
No orbital schedule
Mode JD Channel A
145.850, 145.850
435.910, 435.910
No weekly schedule
No orbital schedule
Mode JD Channel B
145.870, 145.870
435.910, 435.910
No weekly schedule
No orbital schedule
Mode JD Channel C
145.910, 145.910
435.910, 435.910
No weekly schedule
No orbital schedule
end

Weekly and orbital schedules have not yet been implemented in
PREDICT 2.2.0, so for the time being, entries for these parameters
should begin with the word "No".  ALAT/ALON should be entered as
30.0, 0.0 for alat=30.0 and alon=0.0.  See $HOME/.predict/predict.db
for further examples.

Entries in predict.db need not be in the same order as they appear
in predict.tle.  Satellite names need not match either, but the
object numbers in predict.db must be correct.  predict.db may contain
more satellite entries than those appearing in predict.tle (normally 24).
PREDICT reads the entire database file, and reads only the information
that applies to the satellites in the predict.tle file in use at the
time.  If no transponder data is available, or if a satellite contains
no communication transponders, the Single Satellite tracking mode
screen is modified accordingly.

Future revisions of PREDICT will likely include a database editor
as well as possible modifications to the format of the database.


Other changes
=============
PREDICT now installs into a directory whose name includes the revision
number (ie: predict-2.2.0).

Despite the major revisions to the tracking code, all other program
functions, with the exception of the Single Satellite Tracking Mode,
should act as they did in earlier versions.


--
John, KD2BD
October 12, 2002



vocalizer/vocalizer.c
/****************************************************************************
*                  VOCALIZER: A speech articulation program                 *
*		Copyright John A. Magliacane, KD2BD 1999-2003               *
*	   		  Last update: 21-Oct-2015                          *
*****************************************************************************
* This program is used by PREDICT to annunciate satellite coordinates       *
* through the system's soundcard (/dev/dsp) using .wav files containing     *
* uncompressed "raw" PCM data.  It is called as a background process.       *
*****************************************************************************
*                                                                           *
* This program is free software; you can redistribute it and/or modify it   *
* under the terms of the GNU General Public License as published by the     *
* Free Software Foundation; either version 2 of the License or any later    *
* version.                                                                  *
*                                                                           *
* This program is distributed in the hope that it will useful, but WITHOUT  *
* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or     *
* FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License     *
* for more details.                                                         *
*                                                                           *
*****************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <sys/soundcard.h>
#include "vocalizer.h"

int dsp;
char numstr[20][10];
unsigned char buffer[65536];

unsigned long buffer2long(indx)
int indx;
{
	unsigned long byte0, byte1, byte2, byte3;

	if ((indx+3)<65536)
	{
		byte0=(unsigned long)buffer[indx];
		byte1=(unsigned long)buffer[indx+1];
		byte2=(unsigned long)buffer[indx+2];
		byte3=(unsigned long)buffer[indx+3];

		return (byte0|(byte1<<8)|(byte2<<16)|(byte3<<24));
	}
	else
		return 0L;
}

unsigned int buffer2int(indx)
int indx;
{
	unsigned int byte0, byte1;

	if ((indx+1)<65536)
	{
		byte0=(unsigned int)buffer[indx];
		byte1=(unsigned int)buffer[indx+1];

		return (byte0|(byte1<<8));
	}
	else
		return 0;
}

int wavplay(filename)
char *filename;
{
	int x, y, format, bits, bytes, channels, dsp, fd;
	unsigned long total_samples=0L, rate=0L, running_total=0L;
	char filenpath[80];

	strncpy(filenpath,path,79);
	strncat(filenpath,filename,79);
	strncat(filenpath,".wav",79);

	fd=open(filenpath,O_RDONLY);

	if (fd==-1)
	{
		fprintf(stderr,"*** vocalizer: Cannot open \"%s\"\n",filenpath);
		return -1;
	}

	bytes=read(fd,&buffer,65536);

	y=((buffer[0]^'R')  | (buffer[1]^'I') |
  	   (buffer[2]^'F')  | (buffer[3]^'F') |
	   (buffer[8]^'W')  | (buffer[9]^'A') |
	   (buffer[10]^'V') | (buffer[11]^'E'));

	if (y)
	{
		fprintf(stderr,"*** vocalizer: No RIFF header found!\n");
		return -1;
	}

	for (x=12, y=1; x<65535 && y; x++)
		y=((buffer[x]^'f')  | (buffer[x+1]^'m') |
		  (buffer[x+2]^'t') | (buffer[x+3]^' '));

	if (y==0)
	{
		x+=3;
		x+=4;
		format=buffer2int(x);
		x+=2;
		channels=buffer2int(x);
		x+=2;
		rate=buffer2long(x);
		x+=4;
		x+=4;
		/* int block_size=buffer2int(x); */
		buffer2int(x);
		x+=2;
		bits=buffer2int(x);
		x+=2;

		if (format!=1)
		{
			fprintf(stderr,"*** vocalizer: Not PCM Wave Format.\n");
			return -1;
		}

		for (x=0, y=1; x<65535 && y; x++)
			y=((buffer[x]^'d') | (buffer[x+1]^'a') |
			 (buffer[x+2]^'t') | (buffer[x+3]^'a'));

		if (y==0)
		{
			x+=3;
			total_samples=buffer2long(x);

			if (total_samples!=0L)
			{
				dsp=open("/dev/dsp",O_RDWR);

				if (dsp==-1)
				{
					fprintf(stderr,"*** vocalizer: Cannot open /dev/dsp!\n");
					return -1;
				}

				ioctl(dsp,SOUND_PCM_WRITE_CHANNELS,&channels);
				ioctl(dsp,SOUND_PCM_WRITE_RATE,&rate);
				ioctl(dsp,SOUND_PCM_WRITE_BITS,&bits);

				lseek(fd,(long)(x+4),SEEK_SET);
				bytes=read(fd,&buffer,65536);
				running_total+=bytes;

				if (running_total>total_samples)
					bytes-=(running_total-total_samples);

				while (bytes>0)
				{
					write(dsp,&buffer,bytes);

					bytes=read(fd,&buffer,65536);
					running_total+=bytes;

					if (running_total>total_samples)
						bytes-=(running_total-total_samples);
				}

				close(dsp);
				close(fd);
			}
		}
		else
			fprintf(stderr,"*** vocalizer: Could not find 'data' chunk.\n");
	}
	else
		fprintf(stderr,"*** vocalizer: Could not extract format from header.\n");

	return 0;
}

void saynumber(num)
int num;
{
	char string[10];

	sprintf(string,"%4d",num);

	if (string[1]!=32)
	{
		wavplay(numstr[string[1]-48]);
		wavplay("hundred");
	}

	switch (string[2])
	{
		case '1':
			wavplay(numstr[(10+(string[3]-48))]);
			break;

		case '2':	
			wavplay("twenty");
			break;

		case '3':
			wavplay("thirty");
			break;

		case '4':
			wavplay("forty");
			break;

		case '5':
			wavplay("fifty");
			break;

		case '6':
			wavplay("sixty");
			break;

		case '7':
			wavplay("seventy");
			break;

		case '8':
			wavplay("eighty");
			break;

		case '9':
			wavplay("ninety");
			break;
	}

	if (string[3]!=32 && string[3]!=0 && string[2]!='1' && string[3]!='0')
		wavplay(numstr[string[3]-48]);

	if (string[3]=='0' && string[2]==32 && string[1]==32)
		wavplay("zero"); 
}

int main(argc,argv)
int argc;
char *argv[];
{
	int number;

	strcpy(numstr[1], "one\0");
	strcpy(numstr[2], "two\0");
	strcpy(numstr[3], "three\0");
     	strcpy(numstr[4], "four\0");
	strcpy(numstr[5], "five\0");
	strcpy(numstr[6], "six\0");
     	strcpy(numstr[7], "seven\0");
	strcpy(numstr[8], "eight\0");
	strcpy(numstr[9], "nine\0");
     	strcpy(numstr[0], "zero\0");
	strcpy(numstr[10],"ten\0");
	strcpy(numstr[11],"eleven\0");
	strcpy(numstr[12],"twelve\0");
	strcpy(numstr[13],"thirteen\0");
	strcpy(numstr[14],"fourteen\0");
	strcpy(numstr[15],"fifteen\0");
	strcpy(numstr[16],"sixteen\0");
	strcpy(numstr[17],"seventeen\0");
	strcpy(numstr[18],"eighteen\0");
	strcpy(numstr[19],"nineteen\0");

	/* Start talking! */

	if (argc==2)
	{
		if (strncmp(argv[1],"eclipse",7)==0)
		{
			wavplay("alarm");
			wavplay("eclipse");
		}

		if (strncmp(argv[1],"sunlight",8)==0)
		{
			wavplay("alarm");
			wavplay("sunlight");
		}

		if (strncmp(argv[1],"los",3)==0)
			wavplay("los");
	}

	if (argc>=3)
	{
		wavplay("intro");

		number=atoi(argv[1]);
		saynumber(number);
		wavplay("azimuth");

		number=atoi(argv[2]);
		saynumber(number);
		wavplay("elevation");

		if (argc>=4)
		{
			if (argv[3][0]=='+')
				wavplay("approaching");

			if (argv[3][0]=='-')
				wavplay("receding");
		}

		if (argc==5)
		{
			if (argv[4][0]=='V')
				wavplay("visible");
		}
	}

	return 0;
}
